client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,384 @@
+  /**
+   * Allow the passing of an initial value to the resolved promise.
+   * If nothings is passed, the event will be passed to the promise.
+   */
+  let initValue: any;
+  const promise = new Promise<AnimationEvent | TransitionEvent | undefined>(
+    (resolve) => {
+      if (!maxDelay) {
+        resolve(initValue);
+        return;
+      }
+      let timer: number | undefined;
+      const finish = (event: AnimationEvent | TransitionEvent | undefined) => {
+        if (event && event.target !== element) return;
+        window.clearTimeout(timer);
+        element.removeEventListener('transitionend', finish);
+        element.removeEventListener('animationend', finish);
+        resolve(initValue || event);
+      };
+      element.addEventListener('animationend', finish);
+      element.addEventListener('transitionend', finish);
+      timer = window.setTimeout(finish, maxDelay);
+    },
+  );
+  return (init: any) => {
+    initValue = typeof init === 'function' ? init() : init;
+    return promise;
+  };
+};
+
+interface IndexedEventTarget extends EventTarget {
+  index: number;
+}
+
+interface TabLink extends HTMLAnchorElement {
+  index: number;
+}
+
+/**
+ * Adds the ability for the controlled elements to behave as selectable tabs.
+ *
+ * All tabs and tab content must be nested in an element within the scope of the controller.
+ * All tab buttons need the role="tab" attr and an href with the tab content ID with the target 'label'.
+ * Tab contents need to have the role="tabpanel" attribute and and ID attribute that matches the href of the tab link with the target 'panel'.
+ * Tab buttons should also be wrapped in an element with the role="tablist" attribute.
+ * Use the target 'trigger' on an Anchor link and set the href to the #ID of the tab you would like to trigger.
+ *
+ * @example
+ * ```html
+ *      <div data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-selected-class="animate-in">
+ *         <div role="tablist" data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast">
+ *           <a id="tab-label-tab-1" href="#tab-tab-1" role="tab" data-w-tabs-target="label">
+ *             Tab 1
+ *           </a>
+ *           <a id="tab-label-tab-2" href="#tab-tab-2" role="tab" data-w-tabs-target="label">
+ *             Tab 2
+ *           </a>
+ *          </div>
+ *          <div class="tab-content tab-content--comments-enabled">
+ *           <section id="tab-tab-1" role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
+ *             Tab 1 content
+ *           </section>
+ *           <section id="tab-tab-2" role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
+ *             Tab 2 content
+ *           </section>
+ *          </div>
+ *      </div>
+ * ```
+ */
+
```

Comment:
This still needs a bit of refinement, the nesting is not right (2 spaces should be used and aligning opening/closing tags).

Also, a reminder that there should be no line break below the JSDoc and the thing it's describing. See https://jsdoc.app/about-getting-started

```suggestion
 *
 * @example
 * ```html
 * <div data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-selected-class="animate-in">
 *   <div role="tablist" data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast">
 *     <a id="tab-label-tab-1" href="#tab-tab-1" role="tab" data-w-tabs-target="label">Tab 1</a>
 *     <a id="tab-label-tab-2" href="#tab-tab-2" role="tab" data-w-tabs-target="label">Tab 2</a>
 *  </div>
 *    <div class="tab-content tab-content--comments-enabled">
 *     <section id="tab-tab-1" role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
 *       Tab 1 content
 *     </section>
 *     <section id="tab-tab-2" role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
 *       Tab 2 content
 *     </section>
 *   </div>
 * </div>
 * ```
 */
```